gluecodium/src/main/resources/templates/ffi/FfiInstanceCacheImpl.mustache

```diff
@@ -46,18 +49,33 @@ void*
     return reinterpret_cast<std::shared_ptr<void>*>(handle)->get();
 }
 
+// "Private" function, call only when mutex is locked.
+void
+{{libraryName}}_process_deletion_queue(int32_t isolate_id) {
+    auto iter = _deletion_queue.find(isolate_id);
+    if (iter == _deletion_queue.end()) return;
+
+    for (const auto& weak_handle: iter->second) {
+        Dart_DeleteWeakPersistentHandle_DL(weak_handle);
+    }
+    _deletion_queue.erase(iter);
+}
+
 void
-{{libraryName}}_cache_dart_handle_by_raw_pointer(void* raw_pointer, Dart_Handle dart_handle) {
+{{libraryName}}_cache_dart_handle_by_raw_pointer(void* raw_pointer, int32_t isolate_id, Dart_Handle dart_handle) {
     const std::lock_guard<std::mutex> lock(_instance_cache_mutex);
     _instance_cache[raw_pointer] = Dart_NewWeakPersistentHandle_DL(dart_handle, NULL, 0, &{{libraryName}}_nop_finalizer);
+    {{libraryName}}_process_deletion_queue(isolate_id);
 }
 
 void
-{{libraryName}}_uncache_dart_handle_by_raw_pointer(void* raw_pointer) {
+{{libraryName}}_uncache_dart_handle_by_raw_pointer(void* raw_pointer, int32_t isolate_id) {
     const std::lock_guard<std::mutex> lock(_instance_cache_mutex);
     auto iter = _instance_cache.find(raw_pointer);
     if (iter != _instance_cache.end()) {
-        Dart_DeleteWeakPersistentHandle_DL(iter->second);
+        // No _DL API calls can be made if this is called from a finalizer, so weak handle deletion
```

Comment:
Explanation why or reference to a problem is welcome

Reply:
Added a web-link to the official documentation.